### PR TITLE
Fix the unread count not going away when you've signed out

### PR DESCRIPTION
**Pull request by tgolen**
### Fixed Issues
$ https://github.com/Expensify/App/issues/11671

### Tests
**For web, desktop, and mobile web:**
1. Sign in and have another account send you a message
2. Verify that the unread counter goes up (either page title or icon badge). For mobile web, you need to open the browser window which shows the open tabs in order to verify the page titles.
3. Access the unread chat
4. Verify that the unread counter goes down
5. Make sure you have at least 1 unread chat showing on the counter and sign out
6. Verify that the unread counter goes away completely

**For native devices:**
The badge counters don't work in local environment, so they can't really be tested. Just generally do some regression testing to ensure that the unread stuff works like it should

1. Sign in and have another account send you a message
7. Verify that the LHN appears with the unread chat
8. Go into the chat
9. Go back to the LHN and verify the chat is no longer marked as unread

- [ ] Verify that no errors appear in the JS console

### Offline tests
1. Be signed into the app with an unread chat
2. Sign out
3. Verify that the counter goes away

<details>
<summary>web offline test video</summary>

https://user-images.githubusercontent.com/1228807/207709796-5145656b-0604-4ae1-b938-fc2457460a2e.mp4

</details>

### QA Steps
1. Sign in and have another account send you a message
2. Verify that the unread counter goes up (either page title or icon badge). For mobile web, you need to open the browser window which shows the open tabs in order to verify the page titles.
3. Access the unread chat
4. Verify that the unread counter goes down
5. Make sure you have at least 1 unread chat showing on the counter and sign out
6. Verify that the unread counter goes away completely

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/1228807/207711078-dccbea3e-60f7-4ba8-9f38-24aa97dd6aec.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/1228807/207711095-df8df60d-2ea5-4cb1-b04d-4ad39889c396.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/1228807/207711142-fba0e1e9-4ede-45de-b16e-0e7e6783a956.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/1228807/207711165-e882098e-c84d-4f07-8899-759a6487c615.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/1228807/207711194-f2c5d80a-ec06-4235-b4d1-794d12d67e19.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/1228807/207711203-9f588af0-1f19-46b8-a75b-b0af31229c19.mp4

</details>

### CODE REVIEW

**Comment by puneetlath** [comment]
@tgolen lint check is failing.

**Comment by puneetlath** [comment]
@sobitneupane if you're able to get to this soon, then I'll let you take the first crack at it.

**Comment by tgolen** [comment]
Fixed the lint errors :+1:

**Comment by sobitneupane** [comment]
@tgolen I didn't get the expected output. I tried sending some messages. But unread indicator did not update.


https://user-images.githubusercontent.com/25876548/207784602-5e92b7a9-6810-4da1-a24c-d670ec02085c.mov

**Comment by fedirjh** [comment]
> @tgolen I didn't get the expected output. I tried sending some messages. But unread indicator did not update.

@sobitneupane i think that the subscription isn't imported anywhere in the App

File: src/libs/Navigation/AppNavigator/AuthScreens.js [hunk 1]
```diff
@@ -10,7 +10,6 @@ import compose from '../../compose';
 import * as PersonalDetails from '../../actions/PersonalDetails';
 import * as Pusher from '../../Pusher/pusher';
 import PusherConnectionManager from '../../PusherConnectionManager';
-import UnreadIndicatorUpdater from '../../UnreadIndicatorUpdater';
 import ROUTES from '../../../ROUTES';
```
Comment (fedirjh) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:13]:
```suggestion
import '../../UnreadIndicatorUpdater/index';
import ROUTES from '../../../ROUTES';
```

@tgolen I guess the subscription should be imported somewhere in the App

**Review by fedirjh — CHANGES_REQUESTED** [review]

File: src/libs/Navigation/AppNavigator/AuthScreens.js (hunk 1 above)
Comment (tgolen) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:13]: OK, sorry. I think I might have broken this when I just fixed the linter tests via the GitHub web interface. Sorry about that! Let me do this properly and test it out again.

**Comment by tgolen** [comment]
OK, I fixed the lib import and retested it to make sure it's working again.

File: src/libs/Navigation/AppNavigator/AuthScreens.js (hunk 1 above)
Comment (fedirjh) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:13]: changes looks good

**Review by fedirjh — DISMISSED** [review]

**Review by puneetlath — DISMISSED** [review]
It looks good to me, but @tgolen it looks like you have some unit tests failing. 

@sobitneupane will you be able to review/test this PR in the next day?

**Comment by tgolen** [comment]
OK, I'll look into those unit tests.

**Comment by tgolen** [comment]
Unit tests are passing now

**Review by puneetlath — DISMISSED** [review]

**Comment by sobitneupane** [comment]
### Screenshots/Videos
<details>
<summary>Web</summary>

</details>


<details>
<summary>Desktop</summary>


https://user-images.githubusercontent.com/25876548/208104739-f56efa8a-2f42-4fc5-94dc-3b7a04169433.mov


</details>

<details>
<summary>iOS</summary>


https://user-images.githubusercontent.com/25876548/208104795-15c80931-62d5-416f-a032-47216be1df8c.mov


</details>

<details>
<summary>Android</summary>


</details>

**Comment by sobitneupane** [comment]
### Screenshots/Videos
<details>
<summary>Web</summary>

</details>


<details>
<summary>Desktop</summary>


https://user-images.githubusercontent.com/25876548/208104739-f56efa8a-2f42-4fc5-94dc-3b7a04169433.mov


</details>

<details>
<summary>iOS</summary>


https://user-images.githubusercontent.com/25876548/208104795-15c80931-62d5-416f-a032-47216be1df8c.mov


</details>

<details>
<summary>Android</summary>


</details>

**Comment by sobitneupane** [comment]
It tests well on Web. I was unable to test on mWeb due to sign in issues (403 errors). I will test it once my IP will be whitelisted. 

On Desktop, unread indicator in not shown in my device. In fact, It is not shown in staging app as well (so might be some issue on my end)

**Comment by tgolen** [comment]
Could it be your notification settings? @puneetlath Would you be able to test on Desktop and upload a screenshot/video instead?

**Comment by puneetlath** [comment]
Works well for me on Desktop: 

https://user-images.githubusercontent.com/1371996/208196118-85eb5912-03f3-4577-afbd-c9973d176680.mov

**Review by sobitneupane — COMMENTED** [review]
@tgolen I am getting following console error.

Screenshot 2022-12-17 at 10 46 30

**Comment by sobitneupane** [comment]
### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/25876548/208226136-0e27f362-084a-48a1-bead-34417ada6267.mov


</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/25876548/208226182-a22b66ca-6601-4a4f-9e27-517f5ec52fcf.mov


</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/25876548/208226172-30de3196-41a3-4787-8b5d-ebe40cb50503.mov


</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/25876548/208226145-59f084cb-6085-4d84-94aa-3fd0d956ddbe.mov


</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/25876548/208226214-79a97aec-1efc-4638-b990-1b5d7c3720ec.mov


</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/25876548/208226199-40e07559-214f-4836-b035-45fae812e2d2.mov


</details>

**Comment by tgolen** [comment]
> @tgolen I am getting following console error.

OK, I'm not sure how that can happen, but I've added a check for a null report object now which should prevent that from happening.

**Review by puneetlath — APPROVED** [review]

**Comment by puneetlath** [comment]
@sobitneupane are you able to test again?

**Comment by sobitneupane** [comment]
@tgolen With the latest change, when we reload the page after sign out (or close the tab and reopen expensify on new tab after signout), the unread count reappears.

Tested with the following account:
email: applausetester+0901abb@applause.expensifail.com
Password: Feya86Katya

I was able to reproduce the issue on chrome only (not on firefox, safari and chrome incognito).

**Comment by tgolen** [comment]
OK, thanks for reporting that. Let me look into that. That's really strange.

**Comment by tgolen** [comment]
I tried to reproduce it, and I couldn't reproduce it in Chrome on either MacOS or Android mobile web. This is what I did:

1. Sign into an account that has an unread chat
2. Verify the title shows there is an unread chat
3. Sign out
4. Verify the title shows 0 unread chats
5. Refresh
6. Verify the title still shows 0 unread chats

I'm going to remerge `main` into this branch just to see if that helps you. Please make sure to do a fresh `npm i` before testing it again.

**Comment by sobitneupane** [comment]
I was able to reproduce similar issue in staging as well.

When I log out from one account and login to another account, reports from logged out account also appears in the newly logged in account.

So, I think https://github.com/Expensify/App/pull/13600#issuecomment-1358943509 bug is also caused by the same issue.

**Comment by tgolen** [comment]
I don't think I'm following you. The changes in this PR can't affect anything on staging, so the changes here should only improve any problems you saw on staging.

**Comment by sobitneupane** [comment]
I mean to say https://github.com/Expensify/App/pull/13600#issuecomment-1359997799 and https://github.com/Expensify/App/pull/13600#issuecomment-1358943509 bug might be caused by same issue. And might not be related to this PR. I will report it in expensify-bug channel to see if any other user has seen the similar bug.

**Comment by sobitneupane** [comment]
https://user-images.githubusercontent.com/25876548/208750220-0c34fe25-2061-4872-b274-955969a48948.mov

**Comment by puneetlath** [comment]
@sobitneupane so are you still seeing issues with this PR or no? 

I just tried testing on Chrome and wasn't able to reproduce the issue.

**Comment by sobitneupane** [comment]
@puneetlath Yes, I am still reproduce https://github.com/Expensify/App/pull/13600#issuecomment-1358943509 issue. Interestingly, I can only reproduce it on chrome non-incognito.

**Comment by tgolen** [comment]
Hm, I was definitely also testing it in chrom non-icognito. When I first tried to reproduce it, I thought it happened, but then I had some errors in the JS console about missing libs, and that's when I ended up remerging `main` into the branch and restarting webpack. That made the errors go away and I could no longer reproduce it at all.

**Comment by fedirjh** [comment]
@tgolen I haven't tested these changes yet , but what I see in @sobitneupane video I can see where the bug is . i can tell that this is related to both onyx clear logic which delay clearing and the fact that you have imported the subscriber in the whole App . The subsriber should be imported in AuthScreen . Because when he refresh before onyx finish , the data that wasn't cleared will be used by onyx and the subscriber will be triggered again (we have data and subscriber is imported) a fix for this will be to import subscriber on AuthScreen  and/or to close the subscriber connection after logout

**Comment by fedirjh** [comment]
@tgolen as yo can see here the subscriber callback is called on guest screen when Onyx have data (unfinished clearing)

https://user-images.githubusercontent.com/36869046/208905751-92e49910-6e67-4b43-8a1e-aae22213a016.mov

**Comment by tgolen** [comment]
@fedirjh Thanks for those comments! I think it's perfectly fine to have the subscriber where it is. Shouldn't we be more concerned with why there is still report data in Onyx after signing out? That seems to be the root of the issue.

**Comment by tgolen** [comment]
Today I tried to reproduce the issue with a large account. I've got an account with 4000 reports and 4000 users, and I wasn't able to reproduce the issue of Onyx not clearing out all the data. I think we need to try and zero-in on how exactly to reproduce that and treat it as a separate issue. Though, I also saw that we are reverting some of the changes to `Onyx.clear()`, so we might want to hold off on doing anything until we know what the final state of `Onyx.clear()` will be.

For this PR, I think the changes are good and will remain good regardless of what happens to fix `Onyx.clear()`.

**Comment by sobitneupane** [comment]
We have [new issue](https://github.com/Expensify/App/issues/13884) to handle issues like [this](https://github.com/Expensify/App/pull/13600#issuecomment-1358943509). I think we can move forward this this PR.

@puneetlath Overall change looks good to me.

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Pull request merged**